bee-network/bee-autopeering/src/request.rs

```diff
@@ -0,0 +1,202 @@
+// coming in later will be deemed invalid.
+pub(crate) const REQUEST_EXPIRATION_SECS: u64 = 20;
+pub(crate) const EXPIRED_REQUEST_REMOVAL_INTERVAL_CHECK_SECS: u64 = 1;
+pub(crate) const RESPONSE_TIMEOUT: Duration = Duration::from_millis(500);
+
+// Marker trait for requests.
+pub(crate) trait Request: Debug + Clone {}
+
+#[derive(PartialEq, Eq, Hash, Debug, Clone)]
+pub(crate) struct RequestKey {
+    pub(crate) peer_id: PeerId,
+    pub(crate) request_id: TypeId,
+}
+
+pub(crate) struct RequestValue {
+    pub(crate) request_hash: RequestHash,
+    pub(crate) expiration_time: u64,
+    pub(crate) response_tx: Option<ResponseTx>,
+}
+
+#[derive(Clone)]
+pub(crate) struct RequestManager {
+    inner: Arc<RwLock<RequestManagerInner>>,
+}
+
+impl RequestManager {
+    /// Creates a new request manager.
+    pub(crate) fn new(version: u32, network_id: u32, source_addr: SocketAddr) -> Self {
+        Self {
+            inner: Arc::new(RwLock::new(RequestManagerInner {
+                version,
+                network_id,
+                source_addr,
+                open_requests: HashMap::default(),
+            })),
+        }
+    }
+    pub(crate) fn read(&self) -> RwLockReadGuard<RequestManagerInner> {
+        self.inner.read().expect("error getting read access")
+    }
+
+    pub(crate) fn write(&self) -> RwLockWriteGuard<RequestManagerInner> {
+        self.inner.write().expect("error getting write access")
+    }
+}
+
+pub(crate) struct RequestManagerInner {
+    version: u32,
+    network_id: u32,
+    source_addr: SocketAddr,
+    open_requests: HashMap<RequestKey, RequestValue>,
+}
+
+impl RequestManagerInner {
+    pub(crate) fn new_verification_request(
+        &mut self,
+        peer_id: PeerId,
+        peer_addr: IpAddr,
+        response_tx: Option<ResponseTx>,
+    ) -> VerificationRequest {
+        let key = RequestKey {
+            peer_id,
+            request_id: TypeId::of::<VerificationRequest>(),
+        };
+
+        let verif_req = VerificationRequest::new(self.version, self.network_id, self.source_addr, peer_addr);
+        let timestamp = verif_req.timestamp();
+
+        let request_hash = msg_hash(
+            MessageType::VerificationRequest,
+            &verif_req.to_protobuf().expect("error encoding verification request"),
+        );
+
+        let value = RequestValue {
+            request_hash,
+            expiration_time: timestamp + REQUEST_EXPIRATION_SECS,
+            response_tx,
+        };
+
+        let _ = self.open_requests.insert(key, value);
+
+        verif_req
+    }
+
+    pub(crate) fn new_discovery_request(
+        &mut self,
+        peer_id: PeerId,
+        response_tx: Option<ResponseTx>,
+    ) -> DiscoveryRequest {
+        let key = RequestKey {
+            peer_id,
+            request_id: TypeId::of::<DiscoveryRequest>(),
+        };
+
+        let disc_req = DiscoveryRequest::new();
+        let timestamp = disc_req.timestamp();
+
+        let request_hash = msg_hash(
```

Comment:
Would it be an issue if there was a hash collision ? It seems pretty straightforward it's happening but I don't really understand the whole pipeline

Reply:
A hash collision (two actually different byte sequences hashing to the same value) is pretty unlikely for that hash function. Of course that would cause havoc. But I feel like you mean something else. Can you clarify?

Reply:
I meant that the byte sequence is always the same since it's always based on `DiscoveryRequest::new();` but I don't know how this hash is then used and if there is only one DiscoveryRequest at any given time

Reply:
Nah forget about me, I forgot that DiscoveryRequest::new actually creates a timestamp

Reply:
I don't know if that could be an issue but 2 DiscoveryRequest within a second would indeed produce the same hash

Reply:
Right, but even if our node would do that (which it does not), the only thing that would happen is that the "old" DiscoveryRequest would be replaced with the new one (HashMap). Requests are stored by `PeerId` and by "typeof(Request)", so at any point in time you can only have 1 open request type(!!) per peer. That's because it doesn't make sense to have multiple identical requests open at the same time. Of course you can have different requests to the same peer at the same time.